File: examples/06_agent_cards.py
"""Example 06: Agent Profile Discovery via AgentCards

Demonstrates how agents can discover team capabilities through an
AgentCard catalog maintained by the Orchestrator.

Concepts:
- AgentCard: Profile definitions for agent types
- Catalog registration: Adding profiles to the orchestrator
- Discovery: Agents finding profiles by role or skill
- Config: Default config stored in cards
"""

import time

from akgentic.core import (
    ActorSystem,
    AgentCard,
    Akgent,
    BaseConfig,
    Orchestrator,
)

# =============================================================================
# STEP 1: Define Agent Types
# =============================================================================


class ResearchAgent(Akgent):
    """Agent that performs research tasks."""

    def init(self):
        """Initialize the research agent."""
        super().init()
        print(f"[{self.config.name}] Research agent initialized")


class WriterAgent(Akgent):
    """Agent that writes content."""

    def init(self):
        """Initialize the writer agent."""
        super().init()
        print(f"[{self.config.name}] Writer agent initialized")


class CoordinatorAgent(Akgent):
    """Agent that discovers and coordinates other agents."""

    def init(self):
        """Initialize the coordinator agent."""
        super().init()
        print(f"[{self.config.name}] Coordinator initialized")

    def discover_team_capabilities(self) -> None:
        """Discover what agent profiles are available in the team."""
        print(f"\n[{self.config.name}] Discovering team capabilities...")

        # Get full catalog
        catalog = self.discover_catalog()
        print(f"[{self.config.name}] Found {len(catalog)} agent profiles:")
        for card in catalog:
            print(f"  - {card.role}: {card.description}")
            print(f"    Skills: {', '.join(card.skills)}")
            if card.routes_to:
                print(f"    Routes to: {', '.join(card.routes_to)}")
            else:
                print(f"    Routes to: (any role)")

        # Find specific profile
        print(f"\n[{self.config.name}] Looking for ResearchAgent profile...")
        research_profile = self.get_agent_card("ResearchAgent")
        if research_profile:
            print(f"[{self.config.name}] Found ResearchAgent:")
            print(f"  Description: {research_profile.description}")
            print(f"  Skills: {research_profile.skills}")
            config = research_profile.get_config_copy()
            print(f"  Default config: name={config.name}, role={config.role}")

            # Check routing permissions
            print(f"  Can route to WriterAgent: {research_profile.can_route_to('WriterAgent')}")
            print(f"  Can route to AnalystAgent: {research_profile.can_route_to('AnalystAgent')}")
            print(f"  Can route to UnknownAgent: {research_profile.can_route_to('UnknownAgent')}")

        # Find agents with specific skill
        print(f"\n[{self.config.name}] Finding agents with 'writing' skill...")
        writers = self.find_agents_with_skill("writing")
        print(f"[{self.config.name}] Found {len(writers)} agent(s) with writing skill:")
        for card in writers:
            print(f"  - {card.role}")


# =============================================================================
# STEP 2: Main Execution
# =============================================================================


def main() -> None:
    """Run the agent card discovery example."""
    print("[Agent Cards] Starting agent profile discovery demo...\n")

    actor_system = ActorSystem()

    try:
        # Create Orchestrator
        orchestrator_addr = actor_system.createActor(
            Orchestrator,
            config=BaseConfig(name="orchestrator", role="Orchestrator"),
        )

        orch_proxy = actor_system.proxy_ask(orchestrator_addr, Orchestrator)

        # Register agent profiles in the catalog
        print("[Setup] Registering agent profiles in catalog...\n")

        orch_proxy.register_agent_profile(
            AgentCard(
                role="ResearchAgent",
                description="Performs web research and data gathering",
                skills=["web_search", "pdf_extraction", "arxiv_lookup"],
                agent_class="examples.agent_cards.ResearchAgent",
                config=BaseConfig(name="researcher", role="ResearchAgent"),
                routes_to=["WriterAgent", "AnalystAgent"],  # Can only route to these
                metadata={"version": "1.0", "max_concurrent_tasks": 5},
            )
        )

        orch_proxy.register_agent_profile(
            AgentCard(
                role="WriterAgent",
                description="Writes content based on research findings",
                skills=["writing", "summarization", "formatting"],
                agent_class="examples.agent_cards.WriterAgent",
                config=BaseConfig(name="writer", role="WriterAgent"),
                routes_to=["AnalystAgent"],  # Can only route to analyst
                metadata={"version": "1.0", "max_words": 2000},
            )
        )

        orch_proxy.register_agent_profile(
            AgentCard(
                role="AnalystAgent",
                description="Analyzes data and provides insights",
                skills=["data_analysis", "visualization", "web_search"],
                agent_class="examples.agent_cards.AnalystAgent",
                config=BaseConfig(name="analyst", role="AnalystAgent"),
                routes_to=[],  # Empty = no restrictions, can route to anyone
                metadata={"version": "1.0"},
            )
        )

        print("[Setup] Agent profiles registered\n")

        # Query catalog capabilities
        print("[Catalog] Available roles:", orch_proxy.get_available_roles())
        print("[Catalog] Available skills:", orch_proxy.get_available_skills())
        print()

        # Create coordinator agent (via orchestrator to get orchestrator reference)
        coordinator_addr = orch_proxy.createActor(
            CoordinatorAgent,
            config=BaseConfig(name="coordinator", role="CoordinatorAgent"),
        )

        time.sleep(0.2)

        # Coordinator discovers capabilities
        coordinator_proxy = actor_system.proxy_tell(coordinator_addr, CoordinatorAgent)
        coordinator_proxy.discover_team_capabilities()

        time.sleep(0.5)

        print("\n[Agent Cards] Demo complete!")

    finally:
        actor_system.shutdown()


if __name__ == "__main__":
    main()

File: examples/06_agent_cards.exs
# Agent Cards — Capability Discovery and Routing
#
# This example demonstrates the AgentCard catalog pattern:
# - Defining agent profiles with roles, skills, and routing permissions
# - Registering profiles with the Orchestrator catalog
# - Discovering agents by role or skill
# - Checking routing permissions between agents
#
# Maps from Python's AgentCard catalog registration, discovery by role/skill,
# and routing permissions.
#
# Run with: mix run examples/06_agent_cards.exs

alias Akgentic.AgentCard
alias Akgentic.Orchestrator

# =============================================================================
# STEP 1: Create AgentCards describing available agent roles
# =============================================================================

IO.puts("\n=== AgentCard Creation ===\n")

researcher_card =
  AgentCard.new(
    role: "ResearchAgent",
    description: "Researches topics using web search and data extraction",
    skills: ["web_search", "pdf_extraction", "summarization"],
    agent_module: "MyApp.ResearchAgent",
    config: %{name: "researcher", role: "ResearchAgent"},
    routes_to: ["WriterAgent", "AnalystAgent"],
    metadata: %{version: "1.0", max_concurrent_tasks: 3}
  )

writer_card =
  AgentCard.new(
    role: "WriterAgent",
    description: "Writes articles and reports from research findings",
    skills: ["article_writing", "summarization", "formatting"],
    agent_module: "MyApp.WriterAgent",
    config: %{name: "writer", role: "WriterAgent"},
    routes_to: ["ReviewerAgent"],
    metadata: %{version: "1.0"}
  )

analyst_card =
  AgentCard.new(
    role: "AnalystAgent",
    description: "Analyzes data and produces insights",
    skills: ["data_analysis", "visualization", "statistical_modeling"],
    agent_module: "MyApp.AnalystAgent",
    config: %{name: "analyst", role: "AnalystAgent"},
    routes_to: [],
    metadata: %{version: "2.0"}
  )

IO.puts("Created agent cards:")
IO.puts("  - #{researcher_card.role}: #{length(researcher_card.skills)} skills")
IO.puts("  - #{writer_card.role}: #{length(writer_card.skills)} skills")
IO.puts("  - #{analyst_card.role}: #{length(analyst_card.skills)} skills")

# =============================================================================
# STEP 2: Register cards in the Orchestrator catalog
# =============================================================================

IO.puts("\n=== Registering Profiles with Orchestrator ===\n")

{:ok, orch} = Orchestrator.start_link(name: "catalog-orch", timeout_delay: 3600)

Orchestrator.register_agent_profiles(orch, [researcher_card, writer_card, analyst_card])
Process.sleep(100)

catalog = Orchestrator.get_agent_catalog(orch)
IO.puts("Registered #{length(catalog)} agent profiles in the catalog")

# =============================================================================
# STEP 3: Discover agents by role or skill
# =============================================================================

IO.puts("\n=== Discovery: Lookup by Role ===\n")

profile = Orchestrator.get_agent_profile(orch, "WriterAgent")
IO.puts("Found '#{profile.role}': #{profile.description}")
IO.puts("Skills: #{Enum.join(profile.skills, ", ")}")

IO.puts("\n=== Discovery: Lookup by Skill ===\n")

summarizers = Orchestrator.get_profiles_by_skill(orch, "summarization")
IO.puts("Agents with 'summarization' skill:")

Enum.each(summarizers, fn p ->
  IO.puts("  - #{p.role}")
end)

web_searchers = Orchestrator.get_profiles_by_skill(orch, "web_search")
IO.puts("\nAgents with 'web_search' skill:")

Enum.each(web_searchers, fn p ->
  IO.puts("  - #{p.role}")
end)

IO.puts("\n=== Discovery: Available Roles and Skills ===\n")

roles = Orchestrator.get_available_roles(orch)
IO.puts("All available roles: #{Enum.join(roles, ", ")}")

skills = Orchestrator.get_available_skills(orch)
IO.puts("All available skills: #{Enum.join(skills, ", ")}")

# =============================================================================
# STEP 4: Check routing permissions
# =============================================================================

IO.puts("\n=== Routing Permissions ===\n")

IO.puts("ResearchAgent → WriterAgent: #{AgentCard.can_route_to?(researcher_card, "WriterAgent")}")
IO.puts("ResearchAgent → AnalystAgent: #{AgentCard.can_route_to?(researcher_card, "AnalystAgent")}")
IO.puts("ResearchAgent → ReviewerAgent: #{AgentCard.can_route_to?(researcher_card, "ReviewerAgent")}")

IO.puts("WriterAgent → ReviewerAgent: #{AgentCard.can_route_to?(writer_card, "ReviewerAgent")}")
IO.puts("WriterAgent → ResearchAgent: #{AgentCard.can_route_to?(writer_card, "ResearchAgent")}")

# AnalystAgent has empty routes_to — it can route to anyone
IO.puts("AnalystAgent → AnyRole (unrestricted): #{AgentCard.can_route_to?(analyst_card, "AnyRole")}")

# =============================================================================
# STEP 5: Get config copy for safe agent instantiation
# =============================================================================

IO.puts("\n=== Safe Config Copies ===\n")

config = AgentCard.get_config_copy(researcher_card)
IO.puts("Config copy for ResearchAgent: #{inspect(config)}")

module = AgentCard.get_agent_module(researcher_card)
IO.puts("Agent module (from string): #{inspect(module)}")

# Serialize to/from map
map = AgentCard.to_map(writer_card)
IO.puts("\nSerialized WriterAgent card:")
IO.inspect(map, pretty: true)

Orchestrator.stop(orch)

IO.puts("\n[Agent Cards] Done!")
